Consider transitive interfaces when deciding which class to field-inject

mcculls opened the pull request:
Also updates debug logging to only log failed redefines if we would have originally injected the class. This avoids log-spam when we inject the field into a superclass and later on the subclass is redefined.

richardstartin reviewed (commented):
It's better to be more general but I wonder about the cost of doing this, since start up is a pain point. The code wasn't wrong before, it just would have injected more classes than necessary. I'm curious to know if there is a concrete case of a type hierarchy which motivated this, or if this is to make future field injection changes simpler.

mcculls commented:
@richardstartin yes I did find concrete examples where it resulted in unnecessary injection.
 
It can also cause inconsistencies around which types are excluded from injection. Generally if a type is excluded from injection then any of its sub-types are also excluded because they would defer injection to that type. However, if the excluded type indirectly implements the key interface then it will still be excluded (since that's done as a direct name check) but its sub-types could be injected if they directly implement the key interface - because they would be picked up by the old code, while the base type wouldn't. This currently happens for Grizzly's `SafeFutureImpl`, which extends an excluded type but is currently field-injected.

Some performance improvements I have in the pipeline also rely on this static analysis - otherwise we have to do dynamic checks which are slower.

Finally the temporary cache should help reduce costs since it avoids repeated checks, which can happen when you have sub-types re-declaring that they implement the same interfaces - but I'm going to run this against the startup benchmark to get some numbers.

richardstartin reviewed (approved):
All in favour if this is stamping out more edge cases in this feature

mcculls commented:
I've been profiling this change with different applications and so far the addition of the temporary result cache means this PR is often faster than the current code, despite covering more of the hierarchy. For example with the petclinic app it was spending on average 440 microseconds traversing hierarchies in that matcher. With this PR it spends on average 360 microseconds.

I believe this is because there's often repetition in the interface hierarchy.

mcculls commented:
another quick example from the hibernate tests...

before:
```
Added context-store field to org.hibernate.internal.SessionImpl: org.hibernate.SharedSessionContract -> datadog.trace.instrumentation.hibernate.SessionState
Added context-store field to org.hibernate.internal.SessionImpl: org.hibernate.Session -> datadog.trace.instrumentation.hibernate.SessionState
Added context-store field to org.hibernate.internal.SessionImpl: org.hibernate.SharedSessionContract -> datadog.trace.instrumentation.hibernate.SessionState
Added context-store field to org.hibernate.internal.AbstractSessionImpl: org.hibernate.SharedSessionContract -> datadog.trace.instrumentation.hibernate.SessionState
Added context-store field to org.hibernate.internal.AbstractSessionImpl: org.hibernate.Session -> datadog.trace.instrumentation.hibernate.SessionState
Added context-store field to org.hibernate.internal.AbstractSessionImpl: org.hibernate.SharedSessionContract -> datadog.trace.instrumentation.hibernate.SessionState
Added context-store field to org.hibernate.internal.AbstractSharedSessionContract: org.hibernate.SharedSessionContract -> datadog.trace.instrumentation.hibernate.SessionState
Added context-store field to org.hibernate.query.internal.QueryImpl: org.hibernate.Query -> datadog.trace.instrumentation.hibernate.SessionState
Added context-store field to org.hibernate.query.internal.AbstractProducedQuery: org.hibernate.Query -> datadog.trace.instrumentation.hibernate.SessionState
Added context-store field to org.hibernate.engine.transaction.internal.TransactionImpl: org.hibernate.Transaction -> datadog.trace.instrumentation.hibernate.SessionState
```

after:
```
Will not add context-store field to org.hibernate.internal.SessionImpl: org.hibernate.SharedSessionContract -> datadog.trace.instrumentation.hibernate.SessionState, because it will be added to org.hibernate.internal.AbstractSharedSessionContract
Will not add context-store field to org.hibernate.internal.SessionImpl: org.hibernate.Session -> datadog.trace.instrumentation.hibernate.SessionState, because it will be added to org.hibernate.internal.AbstractSessionImpl
Will not add context-store field to org.hibernate.internal.SessionImpl: org.hibernate.SharedSessionContract -> datadog.trace.instrumentation.hibernate.SessionState, because it will be added to org.hibernate.internal.AbstractSharedSessionContract
Will not add context-store field to org.hibernate.internal.AbstractSessionImpl: org.hibernate.SharedSessionContract -> datadog.trace.instrumentation.hibernate.SessionState, because it will be added to org.hibernate.internal.AbstractSharedSessionContract
Added context-store field to org.hibernate.internal.AbstractSessionImpl: org.hibernate.Session -> datadog.trace.instrumentation.hibernate.SessionState
Will not add context-store field to org.hibernate.internal.AbstractSessionImpl: org.hibernate.SharedSessionContract -> datadog.trace.instrumentation.hibernate.SessionState, because it will be added to org.hibernate.internal.AbstractSharedSessionContract
Added context-store field to org.hibernate.internal.AbstractSharedSessionContract: org.hibernate.SharedSessionContract -> datadog.trace.instrumentation.hibernate.SessionState
Will not add context-store field to org.hibernate.query.internal.QueryImpl: org.hibernate.Query -> datadog.trace.instrumentation.hibernate.SessionState, because it will be added to org.hibernate.query.internal.AbstractProducedQuery
Added context-store field to org.hibernate.query.internal.AbstractProducedQuery: org.hibernate.Query -> datadog.trace.instrumentation.hibernate.SessionState
Added context-store field to org.hibernate.engine.transaction.internal.TransactionImpl: org.hibernate.Transaction -> datadog.trace.instrumentation.hibernate.SessionState
```
